packages/components/src/clipboard-button/index.js

```diff
@@ -7,104 +7,77 @@ import classnames from 'classnames';
+	const clipboard = useRef();
+
+	useEffect( () => {
+		let timeoutId;
+
+		clipboard.current = new Clipboard( ref.current, {
+			text: () => ( typeof text === 'function' ? text() : text ),
+			container: ref.current,
 		} );
 
-		this.clipboard.on( 'success', onCopy );
-	}
-
-	componentWillUnmount() {
-		this.clipboard.destroy();
-		delete this.clipboard;
-		clearTimeout( this.onCopyTimeout );
-	}
-
-	onCopy( args ) {
-		// Clearing selection will move focus back to the triggering button,
-		// ensuring that it is not reset to the body, and further that it is
-		// kept within the rendered node.
-		args.clearSelection();
-
-		const { onCopy, onFinishCopy } = this.props;
-		if ( onCopy ) {
-			onCopy();
-			// For convenience and consistency, ClipboardButton offers to call
-			// a secondary callback with delay. This is useful to reset
-			// consumers' state, e.g. to revert a label from "Copied" to
-			// "Copy".
-			if ( onFinishCopy ) {
-				clearTimeout( this.onCopyTimeout );
-				this.onCopyTimeout = setTimeout( onFinishCopy, 4000 );
+		clipboard.current.on( 'success', ( { clearSelection } ) => {
+			// Clearing selection will move focus back to the triggering button,
+			// ensuring that it is not reset to the body, and further that it is
+			// kept within the rendered node.
+			clearSelection();
+
+			if ( onCopy ) {
+				onCopy();
+
+				// For convenience and consistency, ClipboardButton offers to
+				// call a secondary callback with delay. This is useful to reset
+				// consumers' state, e.g. to revert a label from "Copied" to
+				// "Copy".
+				if ( onFinishCopy ) {
+					clearTimeout( timeoutId );
+					timeoutId = setTimeout( onFinishCopy, 4000 );
+				}
 			}
-		}
-	}
-
-	getText() {
-		let text = this.props.text;
-		if ( 'function' === typeof text ) {
-			text = text();
-		}
-
-		return text;
-	}
-
-	render() {
-		const {
-			className,
-			children,
-			// Disable reason: Exclude from spread props passed to Button
-			// eslint-disable-next-line no-unused-vars
-			onCopy,
-			// eslint-disable-next-line no-unused-vars
-			onFinishCopy,
-			// eslint-disable-next-line no-unused-vars
-			text,
-			...buttonProps
-		} = this.props;
-		const classes = classnames( 'components-clipboard-button', className );
+		} );
 
-		// Workaround for inconsistent behavior in Safari, where <textarea> is not
-		// the document.activeElement at the moment when the copy event fires.
-		// This causes documentHasSelection() in the copy-handler component to
-		// mistakenly override the ClipboardButton, and copy a serialized string
-		// of the current block instead.
-		const focusOnCopyEventTarget = ( event ) => {
-			event.target.focus();
+		return () => {
+			clipboard.current.destroy();
+			clearTimeout( timeoutId );
 		};
-
-		return (
-			<Button
-				{ ...buttonProps }
-				className={ classes }
-				ref={ this.containerRef }
-				onCopy={ focusOnCopyEventTarget }
-			>
-				{ children }
-			</Button>
-		);
-	}
+	}, [] );
```

Comment:
there's probably a few things we should add here as deps: onCopy, onFinishCopy too (not sure if I missed something else).

Reply:
Oh yeah, thanks!

Reply:
I wonder if we could create a lint rule to detect the use of variables that are not in the dependency list. Most of the time, they should be added, otherwise we can disable the rule for that case.

Reply:
There's already a lint rule for that, it's disabled in Gutenberg (and there might be an issue to consider enabling it). The problem is that lint rule has impacts. In some situations you want to only run the hook  at mount/unmout even if the prop  changes.  It means you don't want for your component to support the prop  to be changed and that's a totally fine choice in some situations where performance is very important.

For example your change here will affect existing usage. If someone is doing <ClipboardButton onCopy={() => { ) } />` the effect will run on each render. It's not very bad unless that component is in the bottleneck of your UI.

Reply:
Right, but I think it's better to warn by default and make exceptions, either with not including it or memoization. There's less errors if you warn and explicitly need to ignore it.